core/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java

```diff
@@ -394,34 +394,48 @@ public static IndexRoutingTable readFrom(StreamInput in) throws IOException {
          * Initializes a new empty index, as if it was created from an API.
          */
         public Builder initializeAsNew(IndexMetaData indexMetaData) {
-            return initializeEmpty(indexMetaData, true);
+            return initializeEmpty(indexMetaData, true, new UnassignedMeta(UnassignedMeta.Reason.NEW, null));
         }
 
         /**
          * Initializes a new empty index, as if it was created from an API.
          */
         public Builder initializeAsRecovery(IndexMetaData indexMetaData) {
-            return initializeEmpty(indexMetaData, false);
+            return initializeEmpty(indexMetaData, false, new UnassignedMeta(UnassignedMeta.Reason.CLUSTER_RECOVERY, null));
+        }
+
+        /**
+         * Initializes a new empty index, as if it was created from an API.
+         */
+        public Builder initializeAsFromDangling(IndexMetaData indexMetaData) {
+            return initializeEmpty(indexMetaData, false, new UnassignedMeta(UnassignedMeta.Reason.DANGLING_IMPORT, null));
+        }
+
+        /**
+         * Initializes a new empty index, as as a result of opening a closed index.
+         */
+        public Builder initializeAsFromCloseToOpen(IndexMetaData indexMetaData) {
+            return initializeEmpty(indexMetaData, false, new UnassignedMeta(UnassignedMeta.Reason.OPEN, null));
         }
 
         /**
          * Initializes a new empty index, to be restored from a snapshot
          */
         public Builder initializeAsNewRestore(IndexMetaData indexMetaData, RestoreSource restoreSource, IntSet ignoreShards) {
-            return initializeAsRestore(indexMetaData, restoreSource, ignoreShards, true);
+            return initializeAsRestore(indexMetaData, restoreSource, ignoreShards, true, new UnassignedMeta(UnassignedMeta.Reason.NEW_RESTORE, null));
         }
 
         /**
          * Initializes an existing index, to be restored from a snapshot
          */
         public Builder initializeAsRestore(IndexMetaData indexMetaData, RestoreSource restoreSource) {
-            return initializeAsRestore(indexMetaData, restoreSource, null, false);
+            return initializeAsRestore(indexMetaData, restoreSource, null, false, new UnassignedMeta(UnassignedMeta.Reason.EXISTING_RESTORE, null));
```

Comment:
yea, I was thinking about it, and its already part of the cluster state, but might be handy, will add it